### ISSUE TITLE
Change in function massFlowRate corrupts existing models without warning or error

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1973]
**Modified by beutlich on 1 Apr 2016 18:40 UTC**
Let us report on a major issue we detected using Dymola 2017 with MSL3.2.2

In Modelica 3.2.1 the function Fluid.Pipes.Baseclasses.Wallfriction.Partialwallfriction looks like this:

```mo
replaceable partial function massFlowRate_dp 
  "Return mass flow rate m_flow as function of pressure loss dp, i.e., m_flow = f(dp), due to wall friction"
  extends Modelica.Icons.Function;

  input SI.Pressure dp "Pressure loss (dp = port_a.p - port_b.p)";
  input SI.Density rho_a "Density at port_a";
  input SI.Density rho_b "Density at port_b";
  input SI.DynamicViscosity mu_a 
    "Dynamic viscosity at port_a (dummy if use_mu = false)";
  input SI.DynamicViscosity mu_b 
    "Dynamic viscosity at port_b (dummy if use_mu = false)";
  input SI.Length length "Length of pipe";
  input SI.Diameter diameter "Inner (hydraulic) diameter of pipe";
  input SI.Length roughness(min=0) = 2.5e-5 
    "Absolute roughness of pipe, with a default for a smooth steel pipe (dummy if use_roughness = false)";
  input SI.AbsolutePressure dp_small = 1 
    "Regularization of zero flow if |dp| < dp_small (dummy if use_dp_small = false)";
  input SI.ReynoldsNumber Re_turbulent = 4000 
    "Turbulent flow if Re >= Re_turbulent (dummy if use_Re_turbulent = false)";
  output SI.MassFlowRate m_flow "Mass flow rate from port_a to port_b";
annotation (Documentation(info="<html>
</html>"));
end massFlowRate_dp;

```

In Modelica 3.2.2 the function Fluid.Pipes.Baseclasses.Wallfriction.Partialwallfriction looks like this:

```mo
replaceable partial function massFlowRate_dp
  "Return mass flow rate m_flow as function of pressure loss dp, i.e., m_flow = f(dp), due to wall friction"
  extends Modelica.Icons.Function;

  input SI.Pressure dp "Pressure loss (dp = port_a.p - port_b.p)";
  input SI.Density rho_a "Density at port_a";
  input SI.Density rho_b "Density at port_b";
  input SI.DynamicViscosity mu_a
    "Dynamic viscosity at port_a (dummy if use_mu = false)";
  input SI.DynamicViscosity mu_b
    "Dynamic viscosity at port_b (dummy if use_mu = false)";
  input SI.Length length "Length of pipe";
  input SI.Diameter diameter "Inner (hydraulic) diameter of pipe";
  input SI.Area crossArea = pi*diameter^2/4 "Inner cross section area";
  input SI.Length roughness(min=0) = 2.5e-5
    "Absolute roughness of pipe, with a default for a smooth steel pipe (dummy if use_roughness = false)";
  input SI.AbsolutePressure dp_small = 1
    "Regularization of zero flow if |dp| < dp_small (dummy if use_dp_small = false)";
  input SI.ReynoldsNumber Re_turbulent = 4000
    "Turbulent flow if Re >= Re_turbulent (dummy if use_Re_turbulent = false)";
  output SI.MassFlowRate m_flow "Mass flow rate from port_a to port_b";
annotation (Documentation(info="<html>
</html>"));
end massFlowRate_dp;

```

The difference: between diameter and roughness a parameter “crossArea” has been introduced. A typical call of this function looked like this in older models:

```mo
   m_flow[i] = Modelica.Fluid.Pipes.BaseClasses.WallFriction.Detailed.massFlowRate_dp(dp[i], medium[i-1].d, medium[i].d,
   AirMedium.dynamicViscosity(medium[i-1].state), AirMedium.dynamicViscosity(medium[i].state), cellLength, diameter, roughness, 0.01);
```

If this is run with Modelica 3.2.2, the roughness-value is used as crossArea, resulting in a very small calculated mass flow.

In this way existing models are corrupted using Dymola 2017 and MSL3.2.2

Possible corrections:

1) put the new parameter at the end, to simply avoid this issue.

or

2) enforce a better conversion script (i have doubts that this is possible)

or
 
3) use a new function name so that at least this does not happen unnoticed. 

As it is right now, it is quite dangerous since it first happens unnoticed. This issue should be corrected asap. 

----

**Reported by dzimmer on 1 Apr 2016 14:24 UTC**
Let us report on a major issue we detected using Dymola 2017 with MSL3.2.2

In Modelica 3.2.1 the function Fluid.Pipes.Baseclasses.Wallfriction.Partialwallfriction looks like this:

```
replaceable partial function massFlowRate_dp 
  "Return mass flow rate m_flow as function of pressure loss dp, i.e., m_flow = f(dp), due to wall friction"
  extends Modelica.Icons.Function;

  input SI.Pressure dp "Pressure loss (dp = port_a.p - port_b.p)";
  input SI.Density rho_a "Density at port_a";
  input SI.Density rho_b "Density at port_b";
  input SI.DynamicViscosity mu_a 
    "Dynamic viscosity at port_a (dummy if use_mu = false)";
  input SI.DynamicViscosity mu_b 
    "Dynamic viscosity at port_b (dummy if use_mu = false)";
  input SI.Length length "Length of pipe";
  input SI.Diameter diameter "Inner (hydraulic) diameter of pipe";
  input SI.Length roughness(min=0) = 2.5e-5 
    "Absolute roughness of pipe, with a default for a smooth steel pipe (dummy if use_roughness = false)";
  input SI.AbsolutePressure dp_small = 1 
    "Regularization of zero flow if |dp| < dp_small (dummy if use_dp_small = false)";
  input SI.ReynoldsNumber Re_turbulent = 4000 
    "Turbulent flow if Re >= Re_turbulent (dummy if use_Re_turbulent = false)";
  output SI.MassFlowRate m_flow "Mass flow rate from port_a to port_b";
annotation (Documentation(info="<html>
</html>"));
end massFlowRate_dp;

```

In Modelica 3.2.2 the function Fluid.Pipes.Baseclasses.Wallfriction.Partialwallfriction looks like this:

```
replaceable partial function massFlowRate_dp
  "Return mass flow rate m_flow as function of pressure loss dp, i.e., m_flow = f(dp), due to wall friction"
  extends Modelica.Icons.Function;

  input SI.Pressure dp "Pressure loss (dp = port_a.p - port_b.p)";
  input SI.Density rho_a "Density at port_a";
  input SI.Density rho_b "Density at port_b";
  input SI.DynamicViscosity mu_a
    "Dynamic viscosity at port_a (dummy if use_mu = false)";
  input SI.DynamicViscosity mu_b
    "Dynamic viscosity at port_b (dummy if use_mu = false)";
  input SI.Length length "Length of pipe";
  input SI.Diameter diameter "Inner (hydraulic) diameter of pipe";
  input SI.Area crossArea = pi*diameter^2/4 "Inner cross section area";
  input SI.Length roughness(min=0) = 2.5e-5
    "Absolute roughness of pipe, with a default for a smooth steel pipe (dummy if use_roughness = false)";
  input SI.AbsolutePressure dp_small = 1
    "Regularization of zero flow if |dp| < dp_small (dummy if use_dp_small = false)";
  input SI.ReynoldsNumber Re_turbulent = 4000
    "Turbulent flow if Re >= Re_turbulent (dummy if use_Re_turbulent = false)";
  output SI.MassFlowRate m_flow "Mass flow rate from port_a to port_b";
annotation (Documentation(info="<html>
</html>"));
end massFlowRate_dp;

```

The difference: between diameter and roughness a parameter “crossArea” has been introduced. A typical call of this function looked like this in older models:

```
   m_flow[i] = Modelica.Fluid.Pipes.BaseClasses.WallFriction.Detailed.massFlowRate_dp(dp[i], medium[i-1].d, medium[i].d,
   AirMedium.dynamicViscosity(medium[i-1].state), AirMedium.dynamicViscosity(medium[i].state), cellLength, diameter, roughness, 0.01);

```

If this is run with Modelica 3.2.2, the roughness-value is used as crossArea, resulting in a very small calculated mass flow.

In this way existing models are corrupted using Dymola 2017 and MSL3.2.2

Possible corrections:

1) put the new parameter at the end, to simply avoid this issue.

or

2) enforce a better conversion script (i have doubts that this is possible)

or
 
3) use a new function name so that at least this does not happen unnoticed. 

As it is right now, it is quite dangerous since it first happens unnoticed. This issue should be corrected asap. 

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1973

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 1 Apr 2016 18:40 UTC**
See also #1656.

Visual diff:
```diff
--- "a/msl321.mo"
+++ "b/msl322.mo"
@@ -11,6 +11,7 @@ replaceable partial function massFlowRate_dp
     "Dynamic viscosity at port_b (dummy if use_mu = false)";
   input SI.Length length "Length of pipe";
   input SI.Diameter diameter "Inner (hydraulic) diameter of pipe";
+  input SI.Area crossArea = pi*diameter^2/4 "Inner cross section area";
   input SI.Length roughness(min=0) = 2.5e-5
     "Absolute roughness of pipe, with a default for a smooth steel pipe (dummy if use_roughness = false)";
   input SI.AbsolutePressure dp_small = 1
```

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 1 Apr 2016 21:31 UTC**
I looked into the history of this backwards compatibility breaking change and it turns out it was introduced **before** MSL3.2.1+build.3. First in 29a67410d01a196bb2671f3ef4ede06b581ade48 as part of #1601 and then later improved as part of #1656. So this means currently the following versions are **all** broken:

 * `MSL 3.2.1+build.3` released
 * `MSL 3.2.1+build.4` released
 * `MSL 3.2.2+build.2-rc.3` being voted on

Now when looking at #1656 it seems like that **@rfranke** was aware that there was a non-backward compatible change introduced and this on purpose. Only it was never proper mentioned in the release notes. 

So **if** this non-backward change should be kept then I would suggest fixing the release notes of all those versions stating clearly that it is not really a backward compatible change. This would be a documentation fix only. 

If however that should be rolled back then we need at least a `3.2.2+build.3-rc.4` and need to decide later on a `3.2.1+build.5` (we might want to have this anyway).

In addition this shows that we are missing a test model for that function. That should be added as soon as possible.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 1 Apr 2016 21:38 UTC**
I forgot. Of course instead of a roll-back we could fix this like Dirk suggested in this ticket. This at least would mean we have practically only one _broken_ MSL that was shipped with a tool (MSL3.2.1+build.4 if I'm not mistaken).

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 1 Apr 2016 23:37 UTC**
Documented the non-backwards compatible change introduced in version 3.2.1 Build.3 in the release notes with deb769161b7828baeb4c2f4b446b6a11538e8451.

Version 3.2.1 Build.4 is already used in Dymola 2016 FD01 and in OpenModelica. 

I do not see a possibility to fix this issue in a reasonable way. The suggestions from Dirk all have drawbacks:

1. put the new parameter at the end, to simply avoid this issue. -> Models generated with 3.2.1 Build.3 or later will give wrong results.

2. enforce a better conversion script (i have doubts that this is possible) -> conversion scripts can only be provided for a particular version number, but not for a particular build number.

3. use a new function name so that at least this does not happen unnoticed. -> Every fluid model using a pipe model and build with 3.2.1 Build.4 will fail, and the user does not know how to fix this.

After inspecting the code, I wonder how this non-backwards compatible change can give wrong results: These functions are used as building blocks for replaceable packages that are used in the Pipe models to provide different characteristics for wall friction. The functions are hidden in the Pipe models and not directly used by any model in MSL (with exception of the test model Modelica.Fluid.Pipes.BaseClasses.WallFriction.TestWallFrictionAndGravity that was adapted to the argument change). 

Note, there are plenty of test models in Modelica and ModelicaTest where pipe models are used. However, since these functions are never directly called, the test models are still corect. So, this issue cannot be detected by testing.

So, the only way to get a wrong result is by building an own pipe model and using the WallFriction functions.

Probably, proposal (3) is then applicable, so changing the function names (there are 4 functions with this non-backwards compatible change). This will not influence a user that is just using the MSL Pipe models, but will give an error for a user that builds his/her own Pipe models and uses positional function call syntax.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 2 Apr 2016 05:47 UTC**
I guess only rfranke can answer if the breaking change was introduced by accident or by purpose.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 2 Apr 2016 12:55 UTC**
This change was not by accident but intended. The old implementation did only work for circular pipes, even though non-circular pipes can be configured (#1601). Backwards compatibility was improved in #1656. If someone directly uses these base functions with positional arguments for inputs that have default values, then the compilation error actually helps to reconsider if the model at hand is affected by bug #1601.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 2 Apr 2016 17:58 UTC**
I updated the documentation in be54d33ea98d18db935bec2462129285dd2e180f.

Is title of this ticket correct "without warning or error"?
Note that the arguments have different physical units `m` vs. `m2`!

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 3 Apr 2016 09:19 UTC**
I made now a more thorough analysis of the issue, the potential consequences and evaluated this with the attached test package "TestIssue1973":

In version 3.2.1 Build.2 the wall friction model gave wrong results for non-circular pipes (without any message, and even if the user recognizes it, he/she could not correct it). This was fixed by introducing an additional argument crossArea (with a default crossArea for circular pipes) in functions
```
Modelica.Fluid.Pipes.BaseClasses.WallFriction.PartialWallFriction.
    massFlowRate_dp
    massFlowRate_dp_staticHead
    pressureLoss_m_flow
    pressureLoss_m_flow_staticHead
```
The modified DynamicPipe model still gives wrong results, if the new DynamicPipe parameter isCircular is not set to false for a non-circular pipe. However, a user can now set the crossArea for non-circular pipes and can therefore correctly simulate non-circular pipes (which was not possible beforehand).

This bug fix introduced a non-backwards compatible change because a new argument was introduced in the "middle" of the arguments lists of the 4 functions above (with a default value calculated for a circular pipe). This bug fix should have introduced the new argument crossArea at the end of the argument list in order to be fully backwards compatible.

The above change is present in 3.2.1 Build.3, 3.2.1 Build.4, and in 3.2.2 Build.x. 3.2.1 Build.4 is already shipped at least in Dymola 2016 FD01 and in OpenModelica. Since there has been a request to fix the bug for non-circular pipes, it is highly likely that the new feature with the crossArea is already actively used.

From the code review of the library (and evaluated with the test package below), this change has the following consequences on users that generated a corresponding model with MSL 3.2.1 Build.2 and used it with MSl 3.2.1 Build.3 or later:

- If one of the pipe models in Modelica.Fluid.Pipes is used, then everything is fine and no issue occurs, because redeclarations of a WallFriction model requires named arguments (since a modifier), and with a modifier the (correct) default value of the crossArea is automatically used (and in the pipe models the correct function calls are made depending on the used MSL version).

- If a user implements his/her own pipe model, he/she has to call one or more of the 4 functions above and if these functions are called with positional arguments, an error is present, because the new argument crossArea is present. All arguments after crossArea have a different unit as crossArea has. Therefore, tools that check units, will report this issue.

I tested the above statements with the attached test package TestIssue1973 in the following way:

- TestIssue1973.TestFluid1 was generated with MSL 3.2.1 Build.2 (Dymola 2015 FD01) and uses several pipe models where for some pipe models the roughness was changed from the default value of 1e-5 to 1e-9 (the roughess argument is in the argument list after the crossArea argument).

- TestIssue1973.TestFluid1 was also simulated with Dymola 2017 Beta.2 (using MSL 3.2.2 Build.3). No error was reported and the results (stateFM[..]) are identical.

- TestIssue1973.TestFluid2 was generated with MSL 3.2.1 Build.2 (Dymola 2015 FD01) and uses the demo pipe model TestIssue1973.TestWallFrictionAndGravity (a copy of Modelica.Fluid.Pipes.BaseClasses.WallFriction.TestWallFrictionAndGravity that does not have a crossArea argument).

- TestIssue1973.TestFluid2 was also simulated with Dymola 2017 Beta.2 (using MSL 3.2.2 Build.3). A unit error was reported when using the "pedantic" flag set to true, and a unit warning was reported when not using the pedantic flag. If the warning is ignored, Dymola produced a simulation result which is wrong (pipeFriction.V_flow is different in Dymola 2015 FD01 as compared with Dymola 2017 Beta.2).


The question is what do to now. I see the following possibilities (note, in all cases users are not affected that just use the MSL pipe models):

1. Keep 3.2.2 as it is now (with the documentation of this issue in the release notes). The consequence is that if users have build their own pipe models and ignore warnings of the tool with these non-MSL pipe models, then the simulation result will be wrong (but they will get an error, if unit errors stop translation).

2. Rename the WallFriction functions in 3.2.2. The consequence is that every user that has build his/her own pipe model with MSl 3.2.1 Build.2, utilizes it in models and uses these models with MSL 3.2.2 will get an error and has to change the function name and then hopefully regocnices that he/she also has to adapt the function argument list. In case he/she does not recognize that a new argument has been added, the same issue as in (1) will still occur. (Note: If someone made a new pipe model and utilized named arguments for the function calls, his/her model is fine, and yet an error will occur).

3. Move the crossArea argument to the end of the argument lists in 3.2.2 Build.4. The consequence is that every user that has build his/her own pipe model with MSl 3.2.1 Build.3, MSL 3.2.1 Build.4, MSL 3.2.2 Build.3 using function calls with positional arguments and uses a model with this pipe model with MSL 3.2.2 Build.4 or later will get a unit error or warning. If the warning is ignored, the simulation results will be wrong.


My Conclusion is to use variant 1 (keep MSL 3.2.2 as it is now with the improved release notes), because
- in all 3 variants there are scenarios where the user can get wrong simulation results (when ignoring warning messages from the tool),
- a "usual" user is not affected by this change (in any of the variants), because he/she will use the pipe model from MSL. 
- Regression testing of MSL 3.2.1 Build.3, Build.4, MSL 3.2.2 Build.2/.3 have been made with respect to many Modelica libraries, including many fluid libraries. If in these libraries the above issue would be present, there would have been regression errors.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 3 Apr 2016 10:47 UTC**
Just tried `TestIssue1973.TestFluid2` with Dymola 2015FD01 that shipped with MSL3.2.1 build2. When manually loading a newer build of MSL3.2.1 or the trunk version (i.e. MSL3.2.2), this gives the warning:

```
Incompatible units in
pressureLoss_m_flow_Unique10(pipeFriction.m_flow_nominal, pipeFriction.rho_nominal, pipeFriction.rho_nominal, pipeFriction.mu_nominal, pipeFriction.mu_nominal, pipeFriction.length, pipeFriction.diameter, pipeFriction.roughness, pipeFriction.m_flow_small, 0.01, 4000)

The part
crossArea
has unit m2

The part
pipeFriction.roughness
has unit m

The unit issue reported appeared in the binding equation of the variable
pipeFriction.dp_fric_nominal
```

As said: if you directly call these detailed pressure loss functions, then it is worth being noted about #1601, in order to reconsider if your pipe is circular.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 3 Apr 2016 18:16 UTC**
The MA board has discussed this and there is agreement to use variant 1. The release notes have been further improved to make this very clear:
 _In version 3.2.1 Build.3 a new argument crossArea was introduced in the functions of Modelica.Fluid.Pipes.BaseClasses.WallFriction.PartialWallFriction to fix a subtle bug for the calculation of pipe friction for non-circular pipes, see #1601 and #1656. If a user utilized a pipe model of Modelica.Fluid.Pipes, this does not matter because the pipe models have been improved in a fully backwards compatible way. However, if the user constructed an own pipe model based on the partial package PartialWallFriction and calls the functions defined in PartialWallFriction with positional (and not named) arguments, then a unit warning or error will occur (depending on the tool and tool-specific settings) because the new argument crossArea has unit ![m2] and the previous argument at this place, roughness, has unit [m]. If the warning is ignored, the simulation result will be wrong, because the crossArea is used as roughness. The user needs to fix this by adapting his/her pipe model so that the crossArea is used in the function calls, or by using named function arguments._

**Comment by modelica-trac-importer** [comment]
**Comment by fcasella on 3 Apr 2016 21:24 UTC**
Fine for me

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 3 Apr 2016 22:11 UTC**
So is there a way to get a testmodel for the function in ModelicaTest so that future changes might at least trigger some warning/errors?

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 4 Apr 2016 03:05 UTC**
There is:
    ModelicaTest.Fluid.TestComponents.Fittings.TestWallFriction
which uses:
    Modelica.Fluid.Pipes.BaseClasses.WallFriction.TestWallFrictionAndGravity

The latter was adapted along with the fix.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 4 Apr 2016 07:35 UTC**
@dzimmer: after having caused so much trouble right before the release of MSL 3.2.2, can you please answer the two open questions:

1) Do you really don't get a warning, i.e. does you model use physical units?

2) Do you model a circular pipe or possibly a non-circular air duct?

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 4 Apr 2016 08:15 UTC**
@rfranke Could you please open a new ticket since it is crucial that his ticket stays within the 3.2.2 milestone for documentation.

In general I find the "solution" as decided by the board (not MAP-LIB project mind) worrying. 

The function that got changed in the non-backward compatible way has been present in the MSL since `v3.1` (14th August 2009). So this means for around **6 years**. Whereas the newly fixed version with the non-backward compatible change for about 6 **month**. So the likelihood that users will have the old version in their models is **far** higher than the new implementation. So fixing the mistake and appending the argument like in the proposed variant 3, would impact **far less** potential users than keeping the bug in the MSL as it is now. 

So I really don't understand from a user's point of view how this issue has been fixed in the best possible way impacting the least users.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 4 Apr 2016 08:33 UTC**
Ticket #1601 was closed **15 months** ago, after the crossArea argument had been introduced. The leading **1** reduces the likelihood of people being affected now by a factor of 2.5.

Please keep in mind: the crossArea argument is not an extension, which should have been added at the end of the list. It fixes a severe bug. It is considered an advantage if people directly using this base function get notified by the tool.

**Comment by modelica-trac-importer** [comment]
**Comment by dzimmer on 4 Apr 2016 08:44 UTC**
Sorry for causing the trouble... 

To clarify:

We have used pyhsical units (all nice and proper SI units) and got no warning in Dymola whatsoever. We jsut confirmed that once more in order to be sure.

I may also add that industrial clients of us NEVER read the release notes of the MSL. For them it is just a new Dymola version and things suddenly stop working.

**Comment by modelica-trac-importer** [comment]
**Comment by rfranke on 4 Apr 2016 08:50 UTC**
The attached model TestIssue1973.mo generates a warning with default settings of Dymola 2015FD01, see comment:9. What is different in your case?

If people don't read the release notes, then it is even more important that the tool issues a warning or error for the model at hand.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 4 Apr 2016 08:52 UTC**
Replying to [comment:16 rfranke]:
> Ticket #1601 was closed **15 months** ago, after the crossArea argument had been introduced. The leading **1** reduces the likelihood of people being affected now by a factor of 2.5.

I'm only discussing the time line of actual released MSL versions. So in this case 3.2.1 build.4 (build.3 was never really released since some severe issues were found).

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 4 Apr 2016 11:26 UTC**
I am closing this ticket for the documentation of the MSL 3.2.2 release. Please, continue discussion in the new ticket #1974 (with milestone Design90)

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 4 Apr 2016 12:20 UTC**
OK, I can regenerate the ResolvedTracTickets.html now. @otter: can you please fill the release notes field of this ticket. Thanks.

**Comment by modelica-trac-importer** [comment]
**Comment by hubertus on 4 Apr 2016 12:23 UTC**
Replying to [comment:15 dietmarw]:

> The function that got changed in the non-backward compatible way has been present in the MSL since `v3.1` (14th August 2009). So this means for around **6 years**. Whereas the newly fixed version with the non-backward compatible change for about 6 **month**. So the likelihood that users will have the old version in their models is **far** higher than the new implementation. So fixing the mistake and appending the argument like in the proposed variant 3, would impact **far less** potential users than keeping the bug in the MSL as it is now. 
> 
> So I really don't understand from a user's point of view how this issue has been fixed in the best possible way impacting the least users.

Dietmar, I agree with your argument about the number of affected users, based on 6 years vs. 6 months, so I think we should fix the issue as suggested in the ticket by changing the argument order. But I'd like to separate 2 issues here:

 - Our release process for a new release which is quite slow and
 - the proper fixing of a bug that is discovered. 

We have plenty of other, unknown bugs in MSL, there will be more that are going to be discovered soon. 

It would be great if we have a reasonably simple and swift process to fix such bugs, add them to a released version with a new build number, and make it easy for vendors to pick up an official MSL version that includes the fix. Sort of like continuous integration, but for bug fixes only. This should be discussed in map-lib.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 4 Apr 2016 13:01 UTC**
Replying to [comment:22 hubertus]:
> Replying to [comment:15 dietmarw]:
> 
> > The function that got changed in the non-backward compatible way has been present in the MSL since `v3.1` (14th August 2009). So this means for around **6 years**. Whereas the newly fixed version with the non-backward compatible change for about 6 **month**. So the likelihood that users will have the old version in their models is **far** higher than the new implementation. So fixing the mistake and appending the argument like in the proposed variant 3, would impact **far less** potential users than keeping the bug in the MSL as it is now. 
> > 
> > So I really don't understand from a user's point of view how this issue has been fixed in the best possible way impacting the least users.

The argument was not to fix this in a way that impacts the least users, because it is impossible to judge this:
  The user can only get a problem if he/she constructed his/her own pipe model. An industrial user will not do this. So, in nearly all cases, someone else has done this, in most cases it will be a Fluid library developer. However, all important Fluid libraries have been tested against MSL 3.2.1 Build.3, Build.4 and MSL 3.2.2. So the likelyhood that this gives an issue with these libraries is low. The only remaining cases are developments particularly performed for an industrial customer that have not been used for testing against new MSL releases. In case of DLR (that has reported this issue) this is also not critical, because the issue is known now and how to fix it. The industrial customer will just get a new library version and he will be told to use this new library version.

 On the other hand, the request for the bug fix #1601 came from a user that detected that the MSL pipe models cannot be correct for non-circular pipes. The likelyhood is large that this user is already using MSL 3.2.1 Build.4 in order to be able to simulate non-circular pipes (and potentially more users working with non-circular pipes). If this user gets now MSL 3.2.2 that is not backwards compatible to MSL 3.2.1 Build.4, he/she will get trouble again.

Coming back to the start of the discussion: The argument was that any solution will lead to a scenario where some users are potentially unhappy. Having now a strategy: 

1. MSL 3.2.1 Build.4 is not backwards compatible to the previous release MSL 3.2.1 Build.2 (and gives trouble for some users).
2. MSL 3.2.2 is not backwards compatbible to the previous release MSL 3.2.1 Build.4  (and gives trouble for some users).

looks not good at all, and it seems better to have a non-backwards compatible change only once.

Anyway, MSL 3.2.2 is released and we should continue the discussion in #1974.

**Comment by modelica-trac-importer** [comment]
**Comment by hubertus on 4 Apr 2016 13:35 UTC**
I don't htink it is impossible to evaluate how many users are affected, but it is of course an estimate and best guess. We can check what version of what tool used which version of MSL, and make an educated guess. Given how slow commercial users update to the latest version, a change 6 months ago certainly has not propagated a lot.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 4 Apr 2016 14:59 UTC**
Replying to [comment:21 beutlich]:
> @otter: can you please fill the release notes field of this ticket. Thanks.

Reminder.

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 4 Apr 2016 15:07 UTC**

**Comment by modelica-trac-importer** [comment]
**Changelog modified by otter on 4 Apr 2016 15:07 UTC**
Documented the potentially non-backards compatible change in the functions of PartialWallFriction in MSL 3.2.1+Build.3 due to a bug fix regarding non-circular pipes (can be releveant if wall friction functions are directly called, e.g., by pipe models constructed by the user)

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 4 Apr 2016 15:17 UTC**
Replying to [comment:23 otter]:
> Anyway, MSL 3.2.2 is released and we should continue the discussion in #1974.

What do you mean exactly by "released"? Do you mean finalized instead of? Otherwise can we spread the news on modelica.org, twitter etc., let library providers update their libs and finally make a release party?

**Comment by modelica-trac-importer** [comment]
**Comment by mwetter on 4 Apr 2016 15:23 UTC**
Replying to [comment:10 otter]:
> The MA board has discussed this and there is agreement to use variant 1. The release notes have been further improved to make this very clear:
>  _In version 3.2.1 Build.3 a new argument crossArea was introduced in the functions of Modelica.Fluid.Pipes.BaseClasses.WallFriction.PartialWallFriction to fix a subtle bug for the calculation of pipe friction for non-circular pipes, see #1601 and #1656. If a user utilized a pipe model of Modelica.Fluid.Pipes, this does not matter because the pipe models have been improved in a fully backwards compatible way. However, if the user constructed an own pipe model based on the partial package PartialWallFriction and calls the functions defined in PartialWallFriction with positional (and not named) arguments, then a unit warning or error will occur (depending on the tool and tool-specific settings) because the new argument crossArea has unit ![m2] and the previous argument at this place, roughness, has unit [m]. If the warning is ignored, the simulation result will be wrong, because the crossArea is used as roughness. The user needs to fix this by adapting his/her pipe model so that the crossArea is used in the function calls, or by using named function arguments._ 

I am fine with this change.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 5 Apr 2016 08:15 UTC**
Replying to [comment:20 otter]:
> I am closing this ticket for the documentation of the MSL 3.2.2 release.

The documentation was updated by 54cedcfe056668f9edb484c4a006e6ff8ee77686 (trunk) and f1903445bdad3346b131a46648b581cadb884d0a (maint. 3.2.2).
